frequency based negative sampling

AdrianKs opened the pull request:
This is a basic frequency based negative sampling approach.

rgemulla commented:
What's the state of this PR (it's very useful!)?

AdrianKs commented:
It needs to be updated to the new structure of the sampler.
I will update it and let you know.

AdrianKs commented:
The code is ready for review.

rgemulla commented on kge/config-default.yaml:
```diff
@@ -188,7 +188,8 @@ KvsAll:
 negative_sampling:
   # Type of sampler
-  sampling_type: uniform
+  sampling_type: uniform    # uniform, frequency
```
Add a doc string that explains the sampler types.

rgemulla commented on kge/config-default.yaml:
```diff
@@ -188,7 +188,8 @@ KvsAll:
   # Type of sampler
-  sampling_type: uniform
+  sampling_type: uniform    # uniform, frequency
+  symmetric_prior: 1        # this prior is used to smooth counts in frequency sampling
```
Rename to ``frequency.smoothing`` and add a doc string.

rgemulla commented on kge/util/sampler.py:
```diff
@@ -240,3 +241,51 @@ def _filter_and_resample_numba(
+        for slot in SLOTS:
+            vocab = np.arange(self.vocabulary_size[slot].item())
+            unique, counts = np.unique(
+                dataset._triples["train"][:, slot], return_counts=True
```
Use ``dataset.train()``

rgemulla replied:
You can get the counts directly in the format you need using ``np.bincount(y, minlength=<vocab_size>)``.

rgemulla replied:
Then simply add the smoothing value.

rgemulla commented on kge/util/sampler.py:
```diff
@@ -240,3 +241,51 @@ def _filter_and_resample_numba(
+                np.concatenate([counts, np.zeros(len(zero_counts))])[sort_index] + alpha
+            )
+
+            self.samplers.append(
```
self._multinomials

rgemulla commented:
Is it easy to also add the "fast" implementation? (Could also be a separate PR later on.) From my point of view, it may suffice to take this method from KgeUniformSampler and change the line starting with ``new_samples=`` accordingly.

AdrianKs commented:
I addressed your comments.

I think the "fast" implementation is still some work, since in the new_samples= line we directly call the numpy random function.
For frequency sampling we would call the torch multinomial function, which probably is not supported by numba.
When we call the python sample method from the fast numba method it should become slow again.

rgemulla commented:
Thanks!
